tchannel/tornado/dispatch.py

```diff
@@ -233,17 +234,36 @@ def _on_post_response(future):
             e.id = request.id
             connection.send_error(e)
         except Exception as e:
-            error = UnexpectedError(
-                description="Unexpected Error: '%s'" % e,
-                id=request.id,
-                tracing=request.tracing,
-            )
-            response.set_exception(error)
-            connection.request_message_factory.remove_buffer(response.id)
+            try:
+                exc_type, exc_obj, exc_tb = sys.exc_info()
+
+                # Walk to the TB to find our offending line.
+                while exc_tb.tb_next is not None:
+                    exc_tb = exc_tb.tb_next
+
+                description = "%r from %s in %s:%s" % (
+                    e,
+                    request.endpoint,
+                    exc_tb.tb_frame.f_code.co_filename,
+                    exc_tb.tb_lineno,
+                )
+                error = UnexpectedError(
+                    description=description,
+                    id=request.id,
+                    tracing=request.tracing,
+                )
 
-            connection.send_error(error)
-            tchannel.event_emitter.fire(EventType.on_exception, request, error)
-            log.exception(error.description)
+                # Grab exc_info() again because we stomped our tb above.
+                response.set_exception(error, exc_info=sys.exc_info())
+                connection.request_message_factory.remove_buffer(response.id)
+
+                connection.send_error(error)
+                tchannel.event_emitter.fire(EventType.on_exception, request, error)
+                log.error("Unexpected error", exc_info=sys.exc_info())
```

Comment:
You want to use the original output of `sys.exc_info()` here. If any of the functions called above throw or catch exceptions (even as `@coroutine`s), `sys.exc_info()` will be overwritten.